llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -325,23 +325,41 @@ bool RepoPruning::runOnModule(Module &M) {
   }
 
   // Enable this pass support for alias.
-  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
-    Module::alias_iterator CurI = I;
-    ++I;
+  // First, check whether a GlobalAlias references a pruned GO. If yes, record
+  // it. Second, replace all recorded aliases by the declaration. These two
+  // steps can't be merged since the second step could change the base object of
+  // the alias. For example:
+  // Assuming: 1) the aliasee of alias 'A' is alias 'B',
+  //           2) the aliasee of alias 'B' is global object 'C',
+  //           3) the global object 'C' is pruned.
+  // If merging two steps, the order in the alias list is 'B' and then 'A'.
+  // Since the global object 'C' is prund, the alias 'B' is changed from
+  // defination to declaration. Then, since 'B' is changed to declaration, the
+  // base object of the alias 'A' will be changed from 'C' to 'B'. This will
+  // result in the backend error.
+  // Step 1. Record aliases.
+  SmallVector<GlobalAlias *, 1> Aliases;
+  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;
+       ++I) {
     // TODO: An Alias doesn't have an aliasee, for example:
     // @v = alias i32, inttoptr(i32 42 to i32*)
-    if (auto *Aliasee = CurI->getBaseObject()) {
+    if (auto *Aliasee = I->getBaseObject()) {
```

Comment:
Is aliasee the same as baseobject? Your explanation above makes it sound like they're different (particularly in this case). Should the `Aliasee` variable here be `BaseObject`?

Reply:
> 
> 
> Is aliasee the same as baseobject? 

They might be different. 

> Your explanation above makes it sound like they're different (particularly in this case). 

Yes, they are different in the added IR test. 

> Should the `Aliasee` variable here be called `BaseObject`?

Changed, thanks!